framework/include/interfaces/MeshMetaDataInterface.h

```diff
@@ -52,36 +51,109 @@ class MeshMetaDataInterface
    * store. This method will throw an error if the property does not exist.
    */
   template <typename T>
-  const T & getMeshProperty(const std::string & data_name, const std::string & prefix);
+  [[nodiscard]] const T & getMeshProperty(const std::string & data_name,
```

Comment:
I completely disagree. Just as worth it as const. code was removed here as a result of this statement where people were unnecessarily calling this method

Long term - we should use this as much as possible.  This c++ attribute is 6 years old. If we're going to use a modern compiler, we're going to use modern compiler features. Any reviews I do in the future will include requests to add this attribute.

Reply:
@lindsayad, @permcody, @dschwen anyone care to be opinionated here?

Reply:
`[[nodiscard]]` is helpful and and should stay.

Reply:
It's C++ verbiage and its use is very limited (who would get and discard ... and even if you did it by accident it s an obvious one). const serves an actual purpose in restricting API abuse.

Just like it's harder and harder to show an easy kernel to people because of the templating, walking people through function definitions bloated with qualifiers is messier
I m happy with having the discussion here though

Reply:
It's c++ verbiage and we strive to write modern c++ for the modern developer. I think discarding new development practices because we don't want a developer to have to learn new practices is in bad taste

Const serves another purpose: incorrect use of an API, as does an attribute such as this one.

Reply:
I like both https://softwareengineering.stackexchange.com/a/363172 and https://softwareengineering.stackexchange.com/a/363186. From the first link:

> A [[nodiscard]] attribute is most useful in two cases:
>
> 1. if a function has no effects beyond returning a certain result, i.e. is pure. If the result is not used, the call is certainly useless. On the other hand, discarding the result would not be incorrect.
>
> 2. if the return value must be checked, e.g. for a C-like interface that returns error codes instead of throwing. This is the primary use case. For idiomatic C++, that's going to be quite rare.

So for us:

1. As the poster says, a call to a pure function is certainly useless, but it's also not incorrect. As the second link argues, if a user wants do this, do we really care? I suppose that we do if a pure function is also a very expensive one. We probably prefer that users don't do something really expensive for no reason.
2. We do not return error codes

So I think my stance would be, given that `[[nodiscard]]` does indeed clutter the code, we should only mark pure expensive functions as `[[nodiscard]]`

Reply:
Talked to Alex about it some more - I can see the value in not throwing it on _everything_. But I also don't see the value in removing it from things that have had it added

Reply:
> We need them out of here unless you identify an actually expensive pure function where it's defendable

Let's be clear here, this is "I" need them out of here, not "we"

Reply:
If you want to suggest the ones to remove, be my guest. But I will not take the time to go through and remove them. The majority of the ones that are here involve finds/lookups, or other non-free actions.

Reply:
Not removing them all. Quite a few are for factories, have expensive actions that are useless without use, etc. The fact that I _removed_ code that was doing nothing as a result makes the evidence clear. You're worried about the construction of an empty struct that doesn't actually happen but are hell bent on removing attributes that have assisted in this very PR in removing an unnecessary call to building a dependency tree of a significant number of objects?

![image](https://user-images.githubusercontent.com/22062133/221711084-d8850d34-901a-40fa-b67e-a5b71c82e4bf.png)

Please be amenable. I'm trying to find a middle ground here.

Reply:
my views are consistent.
The construction of the empty struct is not a performance problem, I think I ve written that three times already.

You're free to keep the ones that are useful, if any

Reply:
> You're free to keep the ones that are useful, if any

This is not how a review process works - this is the first time I've received an approval. I'm not "free" to do anything, unfortunately. I require your approval.

Reply:
I ll still review it after you've made your selection dont worry.

Reply:
It was pushed before this snarky comment: https://github.com/idaholab/moose/pull/23353/commits/ad30e1f7425c726d9e2eec5474637beaab97d301

Reply:
Verdict?

Reply:
Still too many :O 

it s fine I ll live with it